pkg/net/listener/tls/listener.go

```diff
@@ -0,0 +1,155 @@
+package tls
+
+import (
+	"bufio"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/plgd-dev/hub/v2/pkg/fsnotify"
+	"github.com/plgd-dev/hub/v2/pkg/log"
+	"github.com/plgd-dev/hub/v2/pkg/security/certManager/server"
+)
+
+type ConnectionType uint8
+
+const (
+	ConnectionTypeHTTP ConnectionType = iota
+	ConnectionTypeTLS
+)
+
+type Conn struct {
+	net.Conn
+	ConnectionType ConnectionType
+	buf            *bufio.Reader
+}
+
+func (c *Conn) Read(b []byte) (int, error) {
+	return c.buf.Read(b)
+}
+
+type chConn struct {
+	c   net.Conn
+	err error
+}
+
+type SplitListener struct {
+	net.Listener
+	config *tls.Config
+	cons   chan chConn
+	logger log.Logger
+}
+
+// tlsRecordHeaderLooksLikeHTTP reports whether a TLS record header
+// looks like it might've been a misdirected plaintext HTTP request.
+func tlsRecordHeaderLooksLikeHTTP(hdr []byte) bool {
+	if len(hdr) < 5 {
+		return false
+	}
+	switch string(hdr[:5]) {
+	case "GET /", "HEAD ", "POST ", "PUT /", "OPTIO":
+		return true
+	}
+	return false
+}
+
+func (l *SplitListener) prepareConnection(c net.Conn) {
+	// buffer reads on our conn
+	bconn := &Conn{
+		Conn: c,
+		buf:  bufio.NewReader(c),
+	}
+	n := 5
+	_ = c.SetReadDeadline(time.Now().Add(1 * time.Second))
+	hdr, err := bconn.buf.Peek(n)
+	_ = c.SetReadDeadline(time.Time{})
+	if err != nil {
+		l.logger.Warnf("closing connection for error while peeking for first 5bytes in 1sec: %v", err)
+		_ = bconn.Close()
+		return
+	}
+
+	// I don't remember what the TLS handshake looks like,
+	// but this works as a POC
```

Comment:
is this POC version OK for master, or do we want a better version?

Reply:
Oh this is just copy from Stackoverflow. It is whole reworked :D